From 6104f2b6231747a97d09f818f78d7a9196897005 Mon Sep 17 00:00:00 2001
From: Sven Eckelmann <sven@narfation.org>
Date: Fri, 1 Sep 2023 15:56:45 +0530
Subject: [PATCH] wifi: ath12k: remove call trace while rebooting AP

When a station idles for a long time, hostapd will try to send
a QoS Null frame to the station as "poll". NL80211_CMD_PROBE_CLIENT
is used for this purpose.
And the skb will be added to ack_status_frame - waiting for a
completion via ieee80211_report_ack_skb().

But when the peer was already removed before the tx_complete arrives,
the peer will be missing. And when using dev_kfree_skb_any (instead
of going through mac80211), the entry will stay inside
ack_status_frames thus not clean up related information in its
internal data structures. This IDR will therefore run full after
8K request were generated for such clients.
At this point, the access point will then just stall and not allow
any new clients because idr_alloc() for ack_status_frame will fail.

ieee80211_free_txskb() on the other hand will (when required) call
ieee80211_report_ack_skb() and make sure that (when required) remove
the entry from the ack_status_frame and clean up related
information in its internal data structures.

Signed-off-by: Sven Eckelmann <sven@narfation.org>
Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -882,7 +882,7 @@ ath12k_dp_tx_htt_tx_complete_buf(struct
 	}
 
 	if (unlikely(!skb_cb->vif)) {
-		dev_kfree_skb_any(msdu);
+		ieee80211_free_txskb(ar->ah->hw, msdu);
 		return;
 	}
 
@@ -1131,6 +1131,16 @@ static void ath12k_dp_tx_complete_msdu(s
 
 	flags = skb_cb->flags;
 
+	if (unlikely(!rcu_access_pointer(ab->pdevs_active[ar->pdev_idx]))) {
+                ieee80211_free_txskb(ar->ah->hw, msdu);
+                return;
+        }
+
+        if (unlikely(!skb_cb->vif)) {
+                ieee80211_free_txskb(ar->ah->hw, msdu);
+                return;
+        }
+
 	ath12k_dp_tx_status_parse(ab, tx_status, &ts);
 
 	ar->wmm_stats.tx_type = ath12k_tid_to_ac(ts.tid > ATH12K_DSCP_PRIORITY ? 0:ts.tid);
@@ -1139,16 +1149,6 @@ static void ath12k_dp_tx_complete_msdu(s
 			ar->wmm_stats.total_wmm_tx_drop[ar->wmm_stats.tx_type]++;
 	}
 
-	if (unlikely(!rcu_access_pointer(ab->pdevs_active[ar->pdev_idx]))) {
-		dev_kfree_skb_any(msdu);
-		return;
-	}
-
-	if (unlikely(!skb_cb->vif)) {
-		dev_kfree_skb_any(msdu);
-		return;
-	}
-
 	wbm_status = FIELD_GET(HTT_TX_WBM_COMP_INFO0_STATUS,
 			       tx_status->info0);
 
@@ -1199,7 +1199,7 @@ static void ath12k_dp_tx_complete_msdu(s
 					ath12k_sawf_stats_update(ar, msdu, &ts, peer, timestamp);
 				spin_unlock_bh(&ab->base_lock);
 			}
-			dev_kfree_skb_any(msdu);
+			ieee80211_free_txskb(ar->ah->hw, msdu);
 			return;
 		default:
 			//TODO: Remove this print and add as a stats
@@ -1238,7 +1238,7 @@ static void ath12k_dp_tx_complete_msdu(s
 			   "dp_tx: failed to find the peer with peer_id %d\n",
 			   ts.peer_id);
 		 spin_unlock_bh(&ab->base_lock);
-		 dev_kfree_skb_any(msdu);
+		 ieee80211_free_txskb(ar->ah->hw, msdu);
 		 return;
 	}
 
@@ -1248,7 +1248,7 @@ static void ath12k_dp_tx_complete_msdu(s
 			    peer->addr, peer->peer_id);
 
 		spin_unlock_bh(&ab->base_lock);
-		dev_kfree_skb_any(msdu);
+		ieee80211_free_txskb(ar->ah->hw, msdu);
 		return;
 	}
 
